qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/guava/SmoothRateLimiter.java

```diff
@@ -382,20 +383,31 @@ final long reserveEarliestAvailable(int requiredPermits, long nowMicros) {
     abstract long storedPermitsToWaitTime(double storedPermits, double permitsToTake);
 
     /**
-     * Returns the number of microseconds during cool down that we have to wait to get a new permit.
+     * Returns the number of nanoseconds during cool down that we have to wait to get a new permit.
      */
-    abstract double coolDownIntervalMicros();
+    abstract double coolDownIntervalNanos();
 
     /**
-     * Updates {@code storedPermits} and {@code nextFreeTicketMicros} based on the current time.
+     * Updates {@code storedPermits} and {@code nextFreeTicketNanos} based on the current time.
      */
-    void resync(long nowMicros) {
+    void resync(long nowNanos) {
         // if nextFreeTicket is in the past, resync to now
-        if (nowMicros > nextFreeTicketMicros) {
-            storedPermits = min(maxPermits,
-                    storedPermits
-                            + (nowMicros - nextFreeTicketMicros) / coolDownIntervalMicros());
-            nextFreeTicketMicros = nowMicros;
+        if (nowNanos > nextFreeTicketNanos) {
+            storedPermits = min(maxPermits, storedPermits + (nowNanos - nextFreeTicketNanos) / coolDownIntervalNanos());
+            nextFreeTicketNanos = nowNanos;
+        }
+    }
+
+    void returnPermitsConsumedEarlier(double permitsToReturn, long nowNanos) {
```

Comment:
can we sync in person on how this works?

Reply:
Done.

Reply:
this method should have a changelog